Addressed duplicate methods from reflective lookup

isaki-x opened the pull request:
When implementing interfaces with generic types, in Java 8, the
Class.getMethods() invocation will return methods with matching
signatures but different return types. This fix ensures the most
specific return type is the version chosen.

This fixes the issue brought up in #1123.

qualidafial commented:
Thank you for tracking down and reporting this.

We dealt with a similar issue recently with SQL Object interface types that inherit from generic interfaces. See #1087. 

If my suspicion is correct, this is actually caused by bridge methods (which were introduced way back in Java 5):

```java
interface Foo<T, R> {
  R foo(T input);
}

class Bar implements Foo<String, Integer> {
  Integer foo(String input) {
    // do stuff
  }
}
```

Any time a class/interface extends/implements a type with generic parameters, and provides concrete parameters, you actually end up with two actual methods for each generic method declared:

```java
class Bar implements Foo<String, Integer> {
  // bridge method synthetically created by the Java compiler
  Object foo(Object input) {
    // casts parameters to the expected type and delegates to the other method
    return foo((String) input);
  }

  // generic method
  Integer foo(String input) {
    // do stuff
  }
}
```

Whether the generic method or the bridge method gets called depends on the calling context:

```java
Bar bar = new Bar();
bar.foo("hello"); // calls generic method

Foo<?, ?> foo = bar;
foo.foo("world"); // calls bridge method
```

I think I'd prefer that we fix this by reusing the logic we already have in `BridgeMethodHandlerFactory`. We could introduce a utility method e.g. `GenericTypes.findBridgedMethod(Method bridgeMethod, Class type): Optional<Method>`, and extract the "find bridged method" logic from the factory. Then use that method in both `BridgeMethodHandlerFactory` and `ObjectMethodArguments`.

Finally, we need some tests that reproduce the bug to prove that this change actually fixes the problem, and to prevent future regressions.

isaki-x commented:
Thank you for the feedback! I wrote a quick test mock up to see if that solution and my solution give the same answer (and they did, sort of... they have very different goals; my solution is designed for zero arg methods where the desired method is not known as you are pre-building a cache; the #1087 solution validates method signature matches when the method you want is known and a bridge method and you need the concrete, non-bridged method).

```Java
public class FooTest {

    @Test
    public void test1123() {

        final Map<String, Method> result = Arrays.stream(BlackWidow.class.getMethods())
            .filter(m -> m.getParameterCount() == 0)
            .collect(Collectors.toMap(Method::getName, Function.identity(), FooTest::boundMethodMerge));

        final Method chosen = result.get("getValue");
        assertNotNull(chosen);

        final Class<?> retType = chosen.getReturnType();
        assertEquals(String.class, retType);
    }

    @Test
    public void test1087() {

        // This wants bridge methods for method passed in
        Method tmp = null;
        final Method[] allMethods = BlackWidow.class.getMethods();
        for (int i = 0; i < allMethods.length; ++i) {
            if (allMethods[i].isBridge() && "getValue".equals(allMethods[i].getName())) {
                tmp = allMethods[i];
                break;
            }
        }

        assertNotNull(tmp);

        // Play nice with lambda
        final Method method = tmp;

        final Map<String, Method> result = Arrays.stream(allMethods)
            .filter(candidate -> !candidate.isBridge()
                && Objects.equals(candidate.getName(), method.getName())
                && candidate.getParameterCount() == method.getParameterCount())
            .filter(candidate -> {
                Class<?>[] candidateParamTypes = candidate.getParameterTypes();
                Class<?>[] methodParamTypes = method.getParameterTypes();
                return IntStream.range(0, method.getParameterCount())
                    .allMatch(i -> methodParamTypes[i].isAssignableFrom(candidateParamTypes[i]));
            })
            .collect(Collectors.toMap(Method::getName, Function.identity()));

        final Method chosen = result.get("getValue");
        assertNotNull(chosen);

        final Class<?> retType = chosen.getReturnType();
        assertEquals(String.class, retType);
    }

    private static Method boundMethodMerge(Method a, Method b) {
        final Class<?> aClazz = a.getReturnType();
        final Class<?> bClazz = b.getReturnType();

        final Method ret;
        if (aClazz.isAssignableFrom(bClazz)) {
            ret = b;
        } else if (bClazz.isAssignableFrom(aClazz)) {
            ret = a;
        } else {
            throw new IllegalStateException("Method return conflict");
        }

        return ret;
    }

    private interface Spider<T> {

        T getValue();
    }

    private static final class BlackWidow implements Spider<String> {

        private final String value;

        @SuppressWarnings("unused")
        BlackWidow(final String value) {
            super();
            this.value = value;
        }

        @Override
        public String getValue() {
            return this.value;
        }
    }

}
```

So while these two PRs share a common root cause, the problem they aim to solve is different (as is the information available when tackling the problem).

However, the #1087 solution has given me an idea on how I might do this better. I will write some test cases then come back with an answer as to whether or not I stand by my approach or if I want to change it.

qualidafial commented:
After I left my comment last night, it occurred to me that this might be as simple as making the merge function prefer the non-bridge method over the bridge method.

isaki-x commented:
That is what I'm currently testing.

isaki-x commented:
I found another nuanced issue when developing comprehensive tests. If you have an interface A<T>, a class B<T> that implements A's method that returns T, and then a class C that extends B<T> with a specific type, the specific implementation does not get created.

For example:

```Java
    public interface DatabaseValue<T> {

        T getColumnValue();
    }

    public class DatabaseNumberValue<T extends Number> implements DatabaseValue<T> {

        private final T value;

        DatabaseNumberValue(final T value) {
            super();
            this.value = value;
        }

        @Override
        public T getColumnValue() {
            return this.value;
        }
    }

    public final class DatabaseLongValue extends DatabaseNumberValue<Long> {

        DatabaseLongValue(final long value) {
            super(Long.valueOf(value));
        }
    }

    public final class DatabaseLongValueWithOverride extends DatabaseNumberValue<Long> {

        DatabaseLongValueWithOverride(final long value) {
            super(Long.valueOf(value));
        }

        @Override
        public Long getColumnValue() {
            return super.getColumnValue();
        }
    }

    public final class DirectDatabaseLongValue implements DatabaseValue<Long> {

        private final long value;

        DirectDatabaseLongValue(final long value) {
            super();
            this.value = value;
        }

        @Override
        public Long getColumnValue() {
            return Long.valueOf(this.value);
        }
    }
```

When we dump the methods named 'getColumnValue', we see the following:

`[public java.lang.Number org.jdbi.v3.sqlobject.TestBindMethods$DatabaseNumberValue.getColumnValue(), public java.lang.Object org.jdbi.v3.sqlobject.TestBindMethods$DatabaseNumberValue.getColumnValue()]`

When we explicitly provide an overwrite to the implicitly defined method, you see we get three methods (two of which are bridged):

`[public java.lang.Long org.jdbi.v3.sqlobject.TestBindMethods$DatabaseLongValueWithOverride.getColumnValue(), public java.lang.Number org.jdbi.v3.sqlobject.TestBindMethods$DatabaseLongValueWithOverride.getColumnValue(), public java.lang.Object org.jdbi.v3.sqlobject.TestBindMethods$DatabaseNumberValue.getColumnValue()]`

The reason this is an issue is there is no argument for `Number` so my testing fails.

I'm going to rewrite my tests to use explicit objects that I control so I can verify everything works as desired.

isaki-x commented:
I was wrong; I added an explicit `ArgumentFactory` for `Number` and still get the following error:

```org.jdbi.v3.core.statement.UnableToCreateStatementException: No argument factory registered for type [T] for element [getColumnValue] on [org.jdbi.v3.sqlobject.TestBindMethods$DatabaseLongValue@75881071] [statement:"INSERT INTO bind_methods (id, value) VALUES(:row.getKey, :row.getValue.getColumnValue)", rewritten:"INSERT INTO bind_methods (id, value) VALUES(:row.getKey, :row.getValue.getColumnValue)", parsed:"ParsedSql{sql='INSERT INTO bind_methods (id, value) VALUES(?, ?)', parameters=ParsedParameters{positional=false, parameterNames=[row.getKey, row.getValue.getColumnValue]}}", arguments:{ positional:{}, named:{}, finder:[{lazy object functions arguments "org.jdbi.v3.sqlobject.TestBindMethods$PairRow@365185bd"]}]
```

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBindMethods.java:
```diff
@@ -0,0 +1,253 @@
+
+    }
+
+    public static final class BigIntNumberArgumentFactory implements ArgumentFactory {
```
AbstractArgumentFactory simplifies this process by handling nulls for you--just pass the SQL type integer to the superconstructor and implement the method, which is only invoked when the value is non-null.

isaki-x replied:
After thinking about it, there is no harm in letting the argument factory handle all the cases. I've made your suggested changes and pushed an update. Thanks!

isaki-x commented:
Thanks for both the approval and the compliment!

isaki-x commented:
@qualidafial Is there anything still outstanding that I missed?

qualidafial commented:
Looks great! Thanks for the PR

isaki-x commented:
Sorry to bother you all, but is there an ETA for the next release of JDBI? I'm reaching a deadline for which I need this fix and am trying to manage expectations accordingly. Thanks!

qualidafial commented:
No problem. I'll check with the project members and see if we can cut a release tonight.

isaki-x commented:
That is amazing, thank you!

qualidafial commented:
This is released in 3.3.0. It should be available on Maven Central, although it may not be indexed on search.maven.org for a couple hours.

isaki-x commented:
Thank you so much!
